File: Java/labs/lab20_State/task_3_1 – solution/src/MediaPlayer.java
import java.util.ArrayList;
import java.util.List;

public class MediaPlayer {

    private final List<String> tracks = new ArrayList<>();
    private State state = new StateStopped();
    private int currentTrackNum = 0;


    public String getCurrentTrack() {
        return tracks.get(currentTrackNum);
    }

    public void setTrackNum(int trackNum) {
        if (trackNum < 0 || trackNum >= tracks.size()){
            return;
        }

        this.currentTrackNum = trackNum;
    }

    public int getCurrentTrackNum() {
        return currentTrackNum;
    }

    public List<String> getTracks() {
        return tracks;
    }

    public void addTrack(String track){
        tracks.add(track);
    }

    public void setState(State state) {
        this.state = state;
    }

    public State getState() {
        return state;
    }

    public void play() {
        state.play(this);
    }

    public void pause() {
        state.pause(this);
    }

    public void next(){
        state.next(this);
    }

    public void prev(){
        state.prev(this);
    }

    public void stop() {
        state.stop(this);
    }

}

File: Cpp/lab20_State/task_2_2 - solution/src/media_player.cpp
#include <iostream>

#include "media_player.hpp"


MediaPlayer::~MediaPlayer()
{
  if (this->state != nullptr)
  {
    delete this->state;
    this->state = nullptr;
  }
}

void MediaPlayer::setState(State* state)
{
  if (this->state != nullptr)
  {
    delete this->state;
  }

  this->state = state;
}

State* MediaPlayer::getState() const
{
  return this->state;
}

void MediaPlayer::setIcon(std::string const& icon)
{
  this->icon = icon;
}

std::string const& MediaPlayer::getIcon() const
{
  return this->icon;
}

void MediaPlayer::play()
{
  this->state->play(this);
}

void MediaPlayer::pause()
{
  this->state->pause(this);
}
